src/components/SlateEditor/plugins/id/idPlugin.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright (c) 2025-present, NDLA.
+ *
+ * This source code is licensed under the GPLv3 license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ */
+
+import { createPlugin } from "@ndla/editor";
+import { ID_PLUGIN } from "./idTypes";
+import { Descendant, Editor, Element, Node, NodeEntry } from "slate";
+import { makeNodeId } from "../../utils/makeNodeId";
+
+const assignIdToInitialValue = (editor: Editor, idsSet: Set<string>, nodeEntry: NodeEntry) => {
+  const [node, path] = nodeEntry;
+  if (Element.isElement(node) && editor.isBlock(node)) {
+    if (!node.id && editor.hasPath(path)) {
+      let nodeId = makeNodeId();
+      while (idsSet.has(nodeId)) {
+        nodeId = makeNodeId();
+      }
+      node.id = nodeId;
+    }
+    node.children.forEach((node, idx) => assignIdToInitialValue(editor, idsSet, [node, path.concat(idx)]));
+  }
+};
+
+const getAllIds = (children: Descendant[], set = new Set<string>()): Set<string> => {
+  children.forEach((node) => {
+    if ("id" in node && node.id) {
+      set.add(node.id);
+      node.children?.forEach((child) => getAllIds([child], set));
+    }
+  });
+  return set;
+};
+
+const TRANSFORM_OPS = ["insert_node", "remove_node", "merge_node", "split_node"];
+
+export const idPlugin = createPlugin({
+  name: ID_PLUGIN,
+  normalizeInitialValue: (editor) => {
+    const idsSet = new Set<string>();
+    editor.children.forEach((node, idx) => assignIdToInitialValue(editor, idsSet, [node, [idx]]));
+  },
+  transform: (editor) => {
+    let idsSet = new Set<string>();
+    const { apply } = editor;
+
+    const assignIdRecursively = (node: Node) => {
+      if (Element.isElement(node) && editor.isBlock(node)) {
+        if (!node.id || idsSet.has(node.id)) {
+          let nodeId = makeNodeId();
+          while (idsSet.has(nodeId)) {
+            nodeId = makeNodeId();
+          }
+          idsSet.add(nodeId);
+          node.id = nodeId;
+        }
+        node.children.forEach(assignIdRecursively);
+      }
+    };
+
+    editor.apply = (operation) => {
+      // If there are no ids in the document and our history is empty, the editor is clean. We need to add the initial ids
+      if (
+        (!idsSet.size && (!editor.history.redos.length || !editor.history.undos.length)) ||
+        TRANSFORM_OPS.includes(operation.type)
+      ) {
+        idsSet = getAllIds(editor.children);
```

Comment:
Er det for å sørge for at det ikke er duplikate id'er vi trenger denne?

Reply:
Jepp!

Reply:
Jeg lurer litt på om det egentlig er nødvendig mtp. bruken 🤔  Har du sett denne? https://zelark.github.io/nano-id-cc/

Reply:
Sikkert ikke. Synes du det øker kompleksiteten nok til at vi heller burde fjerne det?

Reply:
Jeg syntes vi godt kan fjerne det!

Reply:
Etter en liten test kom jeg på hvorfor jeg gjorde dette! Slate tar vare på ID'er når man kopiere innhold i en artikkel. Dersom man kopierer to noder og limer de inn igjen vil de derfor bevare ID'ene sine, og da må vi erstatte de på en eller annen måte. Å querye hele dokumentet er garantert tregere enn å sjekke settet vårt.